pkg/dataplane/server/dataplane.go

```diff
@@ -54,24 +59,50 @@ func (d *Dataplane) GetClusterTarget(name string) (string, error) {
 
 // GetClusterHost returns the cluster hostname after trimming ":".
 func (d *Dataplane) GetClusterHost(name string) (string, error) {
+	d.clustersLock.RLock()
+	defer d.clustersLock.RUnlock()
 	if _, ok := d.clusters[name]; !ok {
 		return "", fmt.Errorf("unable to find %s in cluster map", name)
 	}
 	return strings.Split(
 		d.clusters[name].LoadAssignment.GetEndpoints()[0].LbEndpoints[0].GetEndpoint().Hostname, ":")[0], nil
 }
 
-// AddCluster adds a cluster to the map.
+// AddCluster adds/updates a cluster to the map.
 func (d *Dataplane) AddCluster(c *cluster.Cluster) {
+	d.clustersLock.Lock()
+	defer d.clustersLock.Unlock()
 	d.clusters[c.Name] = c
 }
 
+// RemoveCluster adds a cluster to the map.
+func (d *Dataplane) RemoveCluster(name string) {
+	d.clustersLock.Lock()
+	defer d.clustersLock.Unlock()
+	delete(d.clusters, name)
+}
+
+// GetClusters returns the clusters map.
+func (d *Dataplane) GetClusters() map[string]*cluster.Cluster {
+	return d.clusters
+}
+
 // AddListener adds a listener to the map.
 func (d *Dataplane) AddListener(ln *listener.Listener) {
 	listenerName := strings.TrimPrefix(ln.Name, api.ImportListenerPrefix)
-	if _, ok := d.listeners[listenerName]; ok {
-		return
+	d.listenerLock.RLock()
+	if le, ok := d.listeners[listenerName]; ok {
+		// Check if there is an update to the listener address/port
+		if ln.Address.GetSocketAddress().GetAddress() == le.Address.GetSocketAddress().GetAddress() &&
+			ln.Address.GetSocketAddress().GetPortValue() == le.Address.GetSocketAddress().GetPortValue() {
+			d.listenerLock.RUnlock()
+			return
+		}
+		d.listenerEnd[listenerName] <- true
 	}
+	d.listenerLock.RUnlock()
+	d.listenerLock.Lock()
+	defer d.listenerLock.Unlock()
```

Comment:
This seems inefficient to me.
I would take a writer lock at `handleClusters`, and `handleListeners`, and remove all locks from `AddListener`, 'AddCluster`, 'RemoveListener`, `RemoveCluster`.
Just add a comment for each function that says the function assumes lock is already held.